### PR TITLE
fix(addon-editor): prevent erasing history on empty value

### LINKED ISSUE #4290: 🐞 - editor "undo" button cause deletion of all text without able to restore it
Steps to reproduce ([video](https://youtu.be/XvGY680vEvI)):
- Open editor with or without initial value (for example [this page](https://taiga-ui.dev/editor/anchors))
- Press "undo" button (text will disappear)
- Pres "redo" button (nothing happens)

Expected:
- No undo action on initial value
- Redo should work always

### LINKED ISSUE #4293: 🐞 - editor's undo does not work
### Playground Link

https://taiga-ui.tcsbank.ru/editor/API

### Description

Here is video example.
[Taiga-UI-Proprietary-Editor-—-Starter-Kit.webm](https://user-images.githubusercontent.com/53703949/234028473-34d9421c-6ae0-4977-a65c-42bc5d7880e3.webm)

### Angular version

13.3.11

### Taiga UI version

3.25.0

### Which browsers have you used?

- [ ] Chrome
- [ ] Firefox
- [ ] Safari
- [X] Edge

### Which operating systems have you used?

- [ ] macOS
- [X] Windows
- [ ] Linux
- [ ] iOS
- [ ] Android

### LINKED ISSUE #4602: 🐞 - CPU 100% when using large images in editor
### Playground Link

addon-editor

### Description

image

### Angular version

latest

### Taiga UI version

latest

### Which browsers have you used?

- [X] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

### Which operating systems have you used?

- [X] macOS
- [ ] Windows
- [ ] Linux
- [ ] iOS
- [ ] Android

### PR DESCRIPTION
**Pull request by splincode**
## PR Type

What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Refactoring
- [ ] Build or CI related changes
- [ ] Tests related changes
- [ ] Documentation content changes

## What is the current behavior?

Closes #4293 
Closes #4290 
Closes #4602

### CODE REVIEW

File: projects/addon-editor/directives/tiptap-editor/tiptap-editor.directive.ts [hunk 1]
```diff
@@ -26,7 +26,10 @@ import {TuiTiptapEditorService} from './tiptap-editor.service';
 export class TuiTiptapEditorDirective {
     @Input()
     set value(value: string) {
-        this.editor.setValue(value);
```
Comment (waterplea) [inline projects/addon-editor/directives/tiptap-editor/tiptap-editor.directive.ts:29]: Looks like this change should be in the service instead.

File: projects/addon-editor/directives/tiptap-editor/tiptap-editor.service.ts [hunk 2]
```diff
@@ -314,9 +314,17 @@ export class TuiTiptapEditorService extends AbstractTuiEditor {
     }
 
     setValue(value: string): void {
-        if (value !== this.html) {
```
Comment (waterplea) [inline projects/addon-editor/directives/tiptap-editor/tiptap-editor.service.ts:317]: If we move this logic back here, we can skip adding public method of clearing history. It's better not to extend public API unless we have to.

File: projects/addon-editor/directives/tiptap-editor/tiptap-editor.directive.ts (hunk 1 above)
Comment (splincode) [reply projects/addon-editor/directives/tiptap-editor/tiptap-editor.directive.ts]: fixed

File: projects/addon-editor/directives/tiptap-editor/tiptap-editor.service.ts [hunk 3]
```diff
@@ -64,6 +64,23 @@ export class TuiTiptapEditorService extends AbstractTuiEditor {
         });
     }
 
+    @tuiPure
```
Comment (waterplea) [inline projects/addon-editor/directives/tiptap-editor/tiptap-editor.service.ts:67]: Are we sure there is no case when content of the editor was changed some other way and calling `setValue` with the same string as we did before would not be a valid situation to process? Seems like we are only saving up an extra call to `getHTML()` at the cost of uncertainty.

File: projects/addon-editor/directives/tiptap-editor/tiptap-editor.service.ts [hunk 4]
```diff
@@ -64,6 +64,23 @@ export class TuiTiptapEditorService extends AbstractTuiEditor {
...
+    setValue(value: string): void {
+        const html = this.editor.getHTML();
+
+        if (value !== html && !(value === `` && html === `<p></p>`)) {
+            this.editor.commands.setContent(value);
```
Comment (waterplea) [inline projects/addon-editor/directives/tiptap-editor/tiptap-editor.service.ts:72]: Nitpicking: for more than one line we use early return :)

File: projects/addon-editor/directives/tiptap-editor/tiptap-editor.service.ts (hunk 3 above)
Comment (splincode) [reply projects/addon-editor/directives/tiptap-editor/tiptap-editor.service.ts]: fixed

**Review by waterplea — APPROVED** [review]

**Review by vladimirpotekhin — APPROVED** [review]

**Review by waterplea — APPROVED** [review]

**Pull request merged**